### ISSUE TITLE
race condition when registering interfaces - "interface input not found"

### ISSUE THREAD

**Issue by SeanDylanGoff** [directus/directus#10851]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

what i found out so far:

app/src/main.ts calls registerInterfaces(app)

while registerInterfaces(app) is in progress (awaiting the import()'s) the following happens:

- hydrate() is called by onBeforeEach guard in app/src/router.ts
- hydrate calls setLanguage in app/src/lang/set-language.ts which translates interfacesRaw.value and stores the result in interfaces.value

HOWEVER: as registerInterfaces(app) is still in progress this call simply sets interfaces.value to an empty array. this then gets picked up by the interface fallback handling which replaces the actual interface with "input" - and then fails aswell because "input" is missing, too.

i also noticed another problem with probably similar root cause: sometimes the theme "sticks" to light even though the user chose dark. toggling the theme fixes it.

does anyone else have the same problem? a possible fix could be to delay hydration until main.js is fully executed.

### To Reproduce

i don't know how to reliably reproduce this issue - it comes and goes in my dev instance (once it happens it happens over and over again though - even surviving browser cache deletes and vite restarts), never saw it on production.

### Errors Shown

interface "input" not found

### What version of Directus are you using?

9.4.2

### What version of Node.js are you using?

v16.13.1

### What database are you using?

mariadb

### What browser are you using?

chrome

### What operating system are you using?

macOS

### How are you deploying Directus?

custom docker image for development

**Comment by azrikahar** [comment]
That sounds odd, but can't say I've come across this before. 🤔 Though to be fair I don't always have custom interface(s) in my past setups.

Does it only happen to one specific custom interface? Or you've faced this with various different custom interfaces?

Also wondering has this happened in the past and you just so happen to report this now since you're free to do so, or it specifically started happening on `v9.4.2`?

> it comes and goes in my dev instance, never saw it on production.
> custom docker image for development

Would you be able to share the Dockerfile for the custom docker image? May somehow be relevant as to why you're only facing in your said dev instance but not production.

**Comment by SeanDylanGoff** [comment]
> Does it only happen to one specific custom interface? Or you've faced this with various different custom interfaces?
can't really tell as i'm mostly working on one specific interface and can't really force the problem to occur on purpose anyways.
 
> Also wondering has this happened in the past and you just so happen to report this now since you're free to do so, or it specifically started happening on `v9.4.2`?
same problem - a lot of variables changed since i last wrote an interface:
- first time using v9.4.2
- interface is quite the chonker (12 vue files)
- i haven't done a lot of work with this development setup either (the dockerfile/my changes)

my changes (including dockerfile): https://github.com/cynova/directus/commit/63875eeadee519ca54f034143a08e07d09288f8e
image:
https://github.com/cynova/directus/pkgs/container/directus/12578083?tag=dev-9.4.2


> Would you be able to share the Dockerfile for the custom docker image? May somehow be relevant as to why you're only facing in your said dev instance but not production.

however, i think the most probable cause is my physical setup: i'm developing on a remote instance --> more latency. i think this is what might trigger the race condition. 

I added console.log's before and after interface registration, on route setup and inside onBeforeEach hook. 
![image](https://user-images.githubusercontent.com/6235675/148220317-cc625519-7688-4e6b-9bc4-44e8ad019fd0.png)

it's obvious that vue-router does not wait for init().



Possible fix:
i moved app.use(router); after registerInterfaces, panels etc. directly before app.mount.
This seems to solve the problem, and so far i haven't seen any downsides. i'm not sure if that's a clean solution though...

**Comment by rijkvanzanten** [comment]
@azrikahar I think this might be pretty similar to #10707

**Comment by SeanDylanGoff** [comment]
update: i continued working on the interface for long enough that i would expect the problem to occur and it didn't - so moving use(router) really seems to fix it for me. i'll report should this assumption turn out to be false.

**Comment by rijkvanzanten** [comment]
@SeanDylanGoff It makes sense to me to register all the extensions before we tell Vue to fire up, though we'll have to do a bit more testing to see if that has any other unintended side effects (cc @nickrum)

**Comment by nickrum** [comment]
We can't fully register all extensions before initializing Vue though, because most extensions need to globally register components.

**Comment by rijkvanzanten** [comment]
> We can't fully register all extensions before initializing Vue though, because most extensions need to globally register components.

Yeah! But in this case, we'd initialize Vue first, then register the extensions, and then start the router 🤔

**Comment by nickrum** [comment]
I'd rather fix this in a different way. All the App initialization code feels quite unintuitive. But, I don't see why starting the router afterwards wouldn't work 👍

**Comment by rijkvanzanten** [comment]
I'll move this to a discussion so we can keep discussing the various improvements in the loading strategy to optimize for these (edge) cases 🙂
